### PR TITLE
PR: do index analysis only on coordinator

### PR DESCRIPTION
**Pull request by devcrafter**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
During query execution with parallel replicas and enabled local plan, skip index analysis on workers. The coordinator will choose ranges to read for workers based on index analysis on its side (on query initiator).

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 574f26f21fbf5673fcac056d4205c33bacefb893 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72109/574f26f21fbf5673fcac056d4205c33bacefb893/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/353ff951aab88d0231bf12cf72f9d630f800a0e7/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 1]
```diff
@@ -631,10 +633,23 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
     bool use_skip_indexes,
     bool find_exact_ranges)
 {
-    RangesInDataParts parts_with_ranges;
-    parts_with_ranges.resize(parts.size());
     const Settings & settings = context->getSettingsRef();
 
+    if (context->canUseParallelReplicasOnFollower() && settings[Setting::parallel_replicas_local_plan]
```
Comment (nickitat) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:638]: it should work, but there is a caveat. some logic (a.k.a. optimisations) depend on `sum_marks` and if replicas will see much bigger number of marks smth might degrade. not sure how significant that is.

Comment (devcrafter) [reply src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:638]: AFAIS, the `selected_marks` is used only for projection optimisations but they are disabled for queries with parallel replicas - https://github.com/ClickHouse/ClickHouse/blob/cb582abbb51e546a5dc108e76182b6ed57c60b07/src/Processors/QueryPlan/Optimizations/projectionsCommon.cpp#L46-L47

File: src/Core/Settings.cpp [hunk 2]
```diff
@@ -5652,6 +5652,9 @@ Parts virtually divided into segments to be distributed between replicas for par
 )", BETA) \
     DECLARE(Bool, parallel_replicas_local_plan, true, R"(
 Build local plan for local replica
+)", BETA) \
+    DECLARE(Bool, parallel_replicas_skip_index_analysis_on_workers, true, R"(
```
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Core/Settings.cpp:5656]: Well, we skip index analysis only on non-initiator replicas, so just replicas wouldn't be completely correct. In the code we name these nodes followers, - it'd be more consistent. I just don't really like this name and prefer workers :) So, should we stick to followers?

File: src/Core/SettingsChangesHistory.cpp [hunk 3]
```diff
@@ -90,6 +90,7 @@ static std::initializer_list<std::pair<ClickHouseVersion, SettingsChangesHistory
             {"s3_skip_empty_files", false, true, "We hope it will provide better UX"},
             {"filesystem_cache_boundary_alignment", 0, 0, "New setting"},
             {"push_external_roles_in_interserver_queries", false, false, "New setting."},
+            {"parallel_replicas_skip_index_analysis_on_workers", false, true, "Skip index analysis on workers. Effective only with enabled parallel_replicas_local_plan"},
```
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Core/SettingsChangesHistory.cpp:93]: It's already in 24.11

[the remark this answers is not in the document]
Comment (Algunenano) [reply src/Core/SettingsChangesHistory.cpp:93]: 24.12

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp (hunk 1 above)
Comment (nickitat) [reply src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:638]: we also use it here:

https://github.com/ClickHouse/ClickHouse/blob/5e2975baa5a7d45b2f74fbb6b8a1847f2fd1efdc/src/Storages/MergeTree/MergeTreeReadPoolParallelReplicas.cpp#L14-L15

and here:

https://github.com/ClickHouse/ClickHouse/blob/5e2975baa5a7d45b2f74fbb6b8a1847f2fd1efdc/src/Storages/MergeTree/MergeTreeReadPoolBase.cpp#L62-L67

File: src/Core/Settings.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (nickitat) [reply src/Core/Settings.cpp:5652]: follower is a term from master-follower replication scheme, so it is the least favourite for me

[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Core/Settings.cpp]: Renamed to `parallel_replicas_index_analysis_only_on_coordinator`

**Review by nickitat — APPROVED** [review]

**Comment by devcrafter** [comment]
`AST fuzzer (debug) — Logical error: 'This function cannot be used to construct empty tuple. It is a bug'.` - https://github.com/ClickHouse/ClickHouse/issues/72578

**Pull request merged**